server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/cleanup/TransportCleanupRepositoryAction.java

```diff
@@ -0,0 +1,251 @@
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.master.TransportMasterNodeAction;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateUpdateTask;
+import org.elasticsearch.cluster.RepositoryCleanupInProgress;
+import org.elasticsearch.cluster.SnapshotDeletionsInProgress;
+import org.elasticsearch.cluster.SnapshotsInProgress;
+import org.elasticsearch.cluster.block.ClusterBlockException;
+import org.elasticsearch.cluster.block.ClusterBlockLevel;
+import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.repositories.RepositoriesService;
+import org.elasticsearch.repositories.Repository;
+import org.elasticsearch.repositories.RepositoryCleanupResult;
+import org.elasticsearch.repositories.blobstore.BlobStoreRepository;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+
+/**
+ * Repository cleanup action for repository implementations based on {@link BlobStoreRepository}.
+ *
+ * The steps taken by the repository cleanup operation are as follows:
+ * <ol>
+ *     <li>Check that there are no running repository cleanup, snapshot create, or snapshot delete actions
+ *     and add an entry for the repository that is to be cleaned up to {@link RepositoryCleanupInProgress}</li>
+ *     <li>Run cleanup actions on the repository. Note, these are executed exclusively on the master node.
+ *     For the precise operations execute see {@link BlobStoreRepository#cleanup}</li>
+ *     <li>Remove the entry in {@link RepositoryCleanupInProgress} in the first step.</li>
+ * </ol>
+ *
+ * On master failover during the cleanup operation it is simply removed from the cluster state. This is safe because the logic in
+ * {@link BlobStoreRepository#cleanup} ensures that the repository state id has not changed between creation of the cluster state entry
+ * and any delete/write operations. TODO: This will not work if we also want to clean up at the shard level as those will involve writes
+ *                                        as well as deletes.
+ */
+public final class TransportCleanupRepositoryAction extends TransportMasterNodeAction<CleanupRepositoryRequest,
+                                                                                      CleanupRepositoryResponse> {
+
+    private static final Logger logger = LogManager.getLogger(TransportCleanupRepositoryAction.class);
+
+    private static final Version MIN_VERSION = Version.V_8_0_0;
+
+    private final RepositoriesService repositoriesService;
+
+    @Override
+    protected String executor() {
+        return ThreadPool.Names.GENERIC;
```

Comment:
Can't do that, we restrict repository operations to generic and snapshot pool only. Since reading the repo state id is the first step in the master action, it wouldn't add anything to have the management pool in the middle of things here.

Reply:
Interesting, the verify repository action uses the management tread pool but quickly fork to the snapshot one. The get repositories one also declares the management thread pool as executor and also reads the repository metadata. Maybe we should make this more uniform.